src/mol-math/geometry/lookup3d/grid.ts

```diff
@@ -277,4 +289,123 @@ function query<T extends number = number>(ctx: QueryContext, result: Result<T>):
         }
     }
     return result.count > 0;
-}
\ No newline at end of file
+}
+
+const tmpDirVec = Vec3();
+const tmpVec = Vec3();
+const tmpMapG = new Map<number, boolean>();
+const tmpArrG1 = new Array();
+const tmpArrG2 = new Array();
+const tmpNearestIds = new Array();
+function queryNearest<T extends number = number>(ctx: QueryContext, result: Result<T>): boolean {
+    const { expandedBox: box, boundingSphere: { center }, size: [sX, sY, sZ], bucketOffset, bucketCounts, bucketArray, grid, data: { x: px, y: py, z: pz, indices, radius }, delta, maxRadius } = ctx.grid;
+    const [minX, minY, minZ] = box.min;
+    const { x, y, z, k } = ctx;
+    const indicesCount = OrderedSet.size(indices);
+    Result.reset(result);
+    if (indicesCount === 0) return false;
+    let gX, gY, gZ;
+    Vec3.set(tmpVec, x, y, z);
+    if (!Box3D.containsVec3(box, tmpVec)) {
+        // intersect ray pointing to box center
+        Box3D.nearestIntersectionWithRay(tmpVec, box, tmpVec, Vec3.normalize(tmpDirVec, Vec3.sub(tmpDirVec, center, tmpVec)));
+        gX = Math.max(0, Math.min(sX - 1, Math.floor((tmpVec[0] - minX) / delta[0])));
+        gY = Math.max(0, Math.min(sY - 1, Math.floor((tmpVec[1] - minY) / delta[1])));
+        gZ = Math.max(0, Math.min(sZ - 1, Math.floor((tmpVec[2] - minZ) / delta[2])));
+    } else {
+        gX = Math.floor((x - minX) / delta[0]);
+        gY = Math.floor((y - minY) / delta[1]);
+        gZ = Math.floor((z - minZ) / delta[2]);
+    }
+    let gCount = 1, nextGCount = 0, arrG = tmpArrG1, nextArrG = tmpArrG2, prevFurthestDist = Number.MAX_VALUE, prevNearestDist = -Number.MAX_VALUE, nearestDist = Number.MAX_VALUE, findFurthest = true, furthestDist = - Number.MAX_VALUE, distSqG: number;
+    arrG.length = 0;
+    arrG.push(gX, gY, gZ);
+    tmpMapG.clear();
+    while (result.count < k && result.count < indicesCount) {
```

Comment:
This might work, but the implementation feels somewhat weird. There should be a binary/fibonacci heap (or an equivalent priority queue) that keeps the k nearest items.